### ISSUE TITLE
[Feature Request] Scan for IVs and Movesets for certain pokemon

### ISSUE THREAD

**Issue by MangoScango** [RocketMap/RocketMap#1135]
By using the encounter method of the API on discovered pokemon, we can find out all of the pokemon's stats and moveset without even having to catch it. Makes a big difference in letting you know if it's worth chasing down or not. This requires a second API call per pokemon though, so it's best if this is limited to a select few pokemon that the user cares about, Perhaps just put a "get stats" button in each pokemons info box.

Tried to implement this myself, but I'm in a little over my head trying to deal with the frontend and the database. :P

**Comment by Albert62** [comment]
This could work, but if the worker finds the Pokemon, there's a big chance you won't get it with the same CP/IV/Moves.

**Comment by MangoScango** [comment]
From what I understand, IV and moves are predetermined for the encounter,
and don't change from player to player.

**Comment by filippounitn** [comment]
every triner lvl changes iv and cp as far as i know, so if the worker is lvl 1 he will find only  cp and iv only for people lvl 1.
Am i wrong?

**Comment by MangoScango** [comment]
My understanding is that the IVs and moves are the same for every player, but the pokemon's level and therefore CP is dependent on the trainer's level. I'm trying to work on some tests now to verify this.

**Comment by cjhwick** [comment]
IV's, moves, weight, size are fixed and tied to that spawn of that pokemon, everyone who catches that pokemon in the 15min (normally) window, will have the exact same stats, only thing that varies is the level it is (CP/HP), which varies based on trainer level

**Comment by MangoScango** [comment]
Yeah, I just verified this myself. Scanned a Paras with a lvl 1 account, got back:
 `{'move_1': 201, 'move_2': 89, 'pokemon_id': 46, 'height_m': 0.3224717378616333, 'stamina_max': 10, 'weight_kg': 5.651512622833252, 'individual_defense': 3, 'cp_multiplier': 0.09399999678134918, 'stamina': 10, 'individual_stamina': 8, 'individual_attack': 13, 'cp': 11}`

Then went out and caught it with my main, and used GoIV to verify those exact IVs, same height and weight. Is there a list somewhere that I can use to tie these move number to their names? He's got Bug Bite and Cross Poison.

**Comment by NoobDev3** [comment]
I think this would improve the map quite a bit. Also including this information for web hooks would be good. That way alert systems like PokeAlarm could message out based on IV%

**Comment by rrbailey89** [comment]
+1

**Comment by mbg032373** [comment]
I've been pulling in IV information for ALL pokemon on my map for a coule of week now.  I haven't had a single worker get banned the entire time.  You just have to be smart about your searches.

<img width="319" alt="screen shot 2016-09-04 at 7 24 41 pm" src="https://cloud.githubusercontent.com/assets/7226022/18234798/71623166-72d5-11e6-98c3-0e3f6ade6840.png">

**Comment by rrbailey89** [comment]
Also would live a filter for the front-end... Notify if equal to 80% or above..

**Comment by MangoScango** [comment]
@mbg032373 Awesome, would you be willing to submit a PR? Or at least putting your repo up so I can get an idea of what all I'd need to do. I was less worried about bans and more worried about the fact that this will about double the amount of API requests, which is less than ideal no matter bans or no.

**Comment by mbg032373** [comment]
@MangoScango look at PR# 332.  It is what I based mine on.  The folks that run this project already said that they don't want IV information included since it makes the map more like a bot, so there isn't much point in posted a new PR that will just get closed anyway.

That PR is very old and outdated, so there is a bit of work involved in getting it working.  A lot of it is also pretty pointless as well unless you are scanning with workers the same level as your main.  I'm only pulling the Attack, Defense, Stamina and move information.

**Comment by MangoScango** [comment]
@mbg032373 Thanks, I think I'll be able to get that up in my fork, at least. Can definitely understand if this is beyond the scope of the project.

**Comment by djv2** [comment]
@mbg032373 why don't you just put up your repo and link to it (instead of PR)? that way it'll avoid any issues you are mentioning here

**Comment by MangoScango** [comment]
I just pushed a branch I've been working on to my fork if anyone wants to take a look. Still some bugs I'm trying to track down, and it's all pretty messy. But it works as a proof of concept. Definitely don't point your main db at this though.

**Comment by djv2** [comment]
@MangoScango thanks. i see your branch (IVs), but are there specific commits i should be looking at when you began implementing the changes?

**Comment by MangoScango** [comment]
@djv2  d89ae89 is where I merged the pull request, then the next few after that are fixing conflicts. Then it gets messy as I start trying to fix bugs.

**Comment by djv2** [comment]
@MangoScango messy and buggy indeed. stupid question - isn't individual_attack, individual_defense, individual_stamina, move_1, move_2 just a new value that the API returns (along with the old lat, long, encounter_id, time, etc etc)? if yes, then your code seems too complicated? it should be as simple as just adding those values in the array and displaying it in the JS. what am i missing?

**Comment by MangoScango** [comment]
@djv2 A lot of it is just because I'm still new with git and don't really know what I'm doing with it. haha. There's stuff in there that shouldn't be. I'll probably make a cleaner implementation when I get some time. It's not quite that simple though. After the search worker finds a pokemon, you need to parse that info and then make another request to encounter it after, which complicates it a little.  And you have to update the database to hold the new stats.

It's an easier change than my branch makes it look for sure, but more than a couple lines.

**Comment by mbg032373** [comment]
Mine is here:  https://github.com/mbg032373/pogomap/tree/pogowithiv

It is 100% updated with current develop branch here; however, I have changed things that have been merged here that I didn't like.  It also has things hardcoded for my personal use such as the notifications for any pokemon with IV > 0.89.  The notification sound doesn't seem to work on mobile devices either, which I haven't bothered to try to figure out why.

I have also merged a few other PR's from here that have not been merged yet.  For example, you can get pokemon history for each spawnpoint.

In other words, it merges cleanly with develop here, but has several modifications done to suit my own needs.

**Comment by mbg032373** [comment]
@djv2 Your fast with that fork button!  I left some personal info in some of those files.  Can you delete that fork.  I'll repost in a bit without my info in it.

**Comment by djv2** [comment]
@mbg032373 deleted

**Comment by frankalison** [comment]
@mbg032373 Or anyone who knows. :smile: 
how to make my map show the IVs and moves of the Pokemons like you show in https://github.com/PokemonGoMap/PokemonGo-Map/issues/1135#issuecomment-244637386 image?

**Comment by toshioson** [comment]
@MangoScango  Thanks for the great work! Have been trying to pull IV info for some time. One question though, I noticed IV is missing for some pokemon and is it related to the blacklist in your code?

@mbg032373 Do you mind share how to be smart and avoid accounts being banned for pulling IVs? Thanks in advance!

**Comment by djv2** [comment]
@MangoScango - in map.js in function pokemonLabel is where you can call/put pokemon['individual_attack'] etc

**Comment by MangoScango** [comment]
Yeah, the branch in my repo is fully functioning, but I want to reiterate to everyone that it's not really ready to be anyone's actual map. Haha. I still need to implement the blacklist into the config and pull out the wonky changes from my fork.

@toshioson yeah, I've noticed this. The blacklist works properly, I just need to add it to the config instead of hard coding it. I've seen instances of this happening with Pokemon that aren't in the list, and my guess is we're getting rate limited they need a larger encounter delay set

**Comment by EvilLost** [comment]
Just to clarify on some comments earlier.... as mentioned everything is the same for everything except for CP. CP is randomized, *but* it is seeded off of Trainer level. Different trainer levels will see different CPs.

**Comment by djv2** [comment]
@MangoScango was playing with your code this morning, i noticed a lot of times the IV just doesn't show? like for some reason it just decided to skip over scanning IV for a pokemon (even though it should be). thoughts?

**Comment by sebastienvercammen** [comment]
This feature has been added a long time ago. Closing. :)
